File: backend/services/image_analysis.py
import base64
import json
from typing import Any

from backend.clients.gemini import extract_generated_text, post_gemini
from backend.prompts.image_analysis import IMAGE_ANALYSIS_PROMPT


def _parse_json_object(text: str) -> dict[str, Any]:
    cleaned = text.strip()
    if cleaned.startswith("```"):
        lines = cleaned.splitlines()
        if lines and lines[0].startswith("```"):
            lines = lines[1:]
        if lines and lines[-1].strip() == "```":
            lines = lines[:-1]
        cleaned = "\n".join(lines).strip()

    try:
        parsed = json.loads(cleaned)
    except json.JSONDecodeError as exc:
        raise ValueError("Gemini image analysis response was not valid JSON.") from exc
    if not isinstance(parsed, dict):
        raise ValueError("Gemini image analysis response must be a JSON object.")
    return parsed


def _as_string_list(value: Any) -> list[str]:
    if not isinstance(value, list):
        return []
    return [item.strip() for item in value if isinstance(item, str) and item.strip()]


def _normalize_image_analysis(data: dict[str, Any]) -> dict[str, Any]:
    features = data.get("features")
    if not isinstance(features, dict):
        features = data

    normalized_features = {
        "category": str(features.get("category", "")).strip() if features.get("category") else "",
        "colors": _as_string_list(features.get("colors")),
        "materials": _as_string_list(features.get("materials")),
        "style_keywords": _as_string_list(features.get("style_keywords")),
        "use_cases": _as_string_list(features.get("use_cases")),
        "target_audience": _as_string_list(features.get("target_audience")),
        "selling_points": _as_string_list(features.get("selling_points")),
        "detected_text": _as_string_list(features.get("detected_text")),
        "uncertainties": _as_string_list(features.get("uncertainties")),
    }

    recommended_keywords = _as_string_list(data.get("recommendedKeywords"))
    if not recommended_keywords:
        recommended_keywords = [
            keyword
            for keyword in [
                normalized_features["category"],
                *normalized_features["colors"],
                *normalized_features["materials"],
                *normalized_features["style_keywords"],
                *normalized_features["use_cases"],
                *normalized_features["selling_points"],
            ]
            if keyword
        ]

    deduped_keywords = list(dict.fromkeys(recommended_keywords))[:12]
    recommended_summary = str(data.get("recommendedSummary", "")).strip()
    if not recommended_summary:
        summary_parts = [
            normalized_features["category"],
            ", ".join(normalized_features["colors"]),
            ", ".join(normalized_features["style_keywords"] or normalized_features["selling_points"]),
        ]
        recommended_summary = " / ".join(part for part in summary_parts if part)

    return {
        "recommendedKeywords": deduped_keywords,
        "recommendedSummary": recommended_summary,
        "features": normalized_features,
    }


def analyze_product_image(
    image_bytes: bytes,
    media_type: str,
    api_key_override: str | None = None,
) -> dict[str, Any]:
    if not image_bytes:
        raise ValueError("Uploaded image is empty.")

    payload = {
        "contents": [
            {
                "parts": [
                    {"text": IMAGE_ANALYSIS_PROMPT},
                    {
                        "inline_data": {
                            "mime_type": media_type,
                            "data": base64.b64encode(image_bytes).decode("ascii"),
                        }
                    },
                ]
            }
        ],
        "generationConfig": {
            "response_mime_type": "application/json",
        },
    }

    response = (
        post_gemini(payload, image_analysis=True, api_key_override=api_key_override)
        if api_key_override
        else post_gemini(payload, image_analysis=True)
    )
    generated_text = extract_generated_text(response)
    return _normalize_image_analysis(_parse_json_object(generated_text))

File: src/server/ai/imageAnalysis.ts
import { extractGeneratedText, postGemini } from "./geminiClient";
import { IMAGE_ANALYSIS_PROMPT } from "./prompts/imageAnalysis";

export type ImageAnalysisResult = {
  recommendedKeywords: string[];
  recommendedSummary: string;
  features: {
    category: string;
    colors: string[];
    materials: string[];
    style_keywords: string[];
    use_cases: string[];
    target_audience: string[];
    selling_points: string[];
    detected_text: string[];
    uncertainties: string[];
  };
};

function parseJsonObject(text: string): Record<string, unknown> {
  let cleaned = text.trim();
  if (cleaned.startsWith("```")) {
    const lines = cleaned.split(/\r?\n/u);
    if (lines[0]?.startsWith("```")) {
      lines.shift();
    }
    if (lines[lines.length - 1]?.trim() === "```") {
      lines.pop();
    }
    cleaned = lines.join("\n").trim();
  }

  const parsed = JSON.parse(cleaned) as unknown;
  if (!parsed || typeof parsed !== "object" || Array.isArray(parsed)) {
    throw new Error("Gemini image analysis response must be a JSON object.");
  }
  return parsed as Record<string, unknown>;
}

function asStringList(value: unknown): string[] {
  if (!Array.isArray(value)) {
    return [];
  }
  return value.map((item) => (typeof item === "string" ? item.trim() : "")).filter(Boolean);
}

function normalizeImageAnalysis(data: Record<string, unknown>): ImageAnalysisResult {
  const rawFeatures = data.features && typeof data.features === "object" ? (data.features as Record<string, unknown>) : data;
  const features = {
    category: rawFeatures.category ? String(rawFeatures.category).trim() : "",
    colors: asStringList(rawFeatures.colors),
    materials: asStringList(rawFeatures.materials),
    style_keywords: asStringList(rawFeatures.style_keywords),
    use_cases: asStringList(rawFeatures.use_cases),
    target_audience: asStringList(rawFeatures.target_audience),
    selling_points: asStringList(rawFeatures.selling_points),
    detected_text: asStringList(rawFeatures.detected_text),
    uncertainties: asStringList(rawFeatures.uncertainties),
  };

  let recommendedKeywords = asStringList(data.recommendedKeywords);
  if (!recommendedKeywords.length) {
    recommendedKeywords = [
      features.category,
      ...features.colors,
      ...features.materials,
      ...features.style_keywords,
      ...features.use_cases,
      ...features.selling_points,
    ].filter(Boolean);
  }

  let recommendedSummary = data.recommendedSummary ? String(data.recommendedSummary).trim() : "";
  if (!recommendedSummary) {
    recommendedSummary = [features.category, features.colors.join(", "), (features.style_keywords.length ? features.style_keywords : features.selling_points).join(", ")]
      .filter(Boolean)
      .join(" / ");
  }

  return {
    recommendedKeywords: Array.from(new Set(recommendedKeywords)).slice(0, 12),
    recommendedSummary,
    features,
  };
}

export async function analyzeProductImage(params: {
  imageBytes: Buffer;
  mediaType: string;
  apiKeyOverride: string;
}): Promise<ImageAnalysisResult> {
  if (!params.imageBytes.length) {
    throw new Error("Uploaded image is empty.");
  }

  const payload = {
    contents: [
      {
        parts: [
          { text: IMAGE_ANALYSIS_PROMPT },
          {
            inline_data: {
              mime_type: params.mediaType,
              data: params.imageBytes.toString("base64"),
            },
          },
        ],
      },
    ],
    generationConfig: {
      response_mime_type: "application/json",
    },
  };

  const response = await postGemini(payload, {
    imageAnalysis: true,
    apiKeyOverride: params.apiKeyOverride,
  });
  return normalizeImageAnalysis(parseJsonObject(extractGeneratedText(response)));
}
